docs/rfcs/rfc-035-authorization.md

```diff
@@ -0,0 +1,759 @@
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+Given the following GraphQL query:
+
+```gql
+{
+  users(where: { name: "Bob" }) {
+    id
+    name
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.id = $jwt.sub
+AND this.name = $name
+RETURN this { .id, .name }
+```
+
+```json
+{
+  "jwt": {
+    "sub": "123456"
+  },
+  "name": "Bob"
+}
+```
+
+### Validate
+
+Validate rules apply filters also, but throw an error in the case of reaching any data that breaches the predicate. The rules are applied using `apoc.util.validatePredicate`.
+
+By default, a validate rule is applied both before and after every operation. For example, if validating that a `User` node matches the provided JWT by checking the ID matches the subject, the odds are that you are going to want to check this is still the case following the operation.
+
+So given the following example:
+
+```gql
+type User
+  @authorization(
+    validate: [{ where: { node: { id: "$jwt.sub" } } }]
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+And the following GraphQL query:
+
+```gql
+mutation {
+  updateUsers(where: { name: "Bob" }, update: { id: "654321" }) {
+    users {
+      id
+      name
+    }
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.name = $name
+AND apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])
+SET this.id = $update.id
+WITH this
+WHERE apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])
+RETURN this { .id, .name }
+```
+
+```json
+{
+  "jwt": {
+    "sub": "123456"
+  },
+  "name": "Bob",
+  "update": {
+    "id": "654321"
+  }
+}
+```
+
+If, following the Mutation operation, the rule is not satisfied, an error "Unauthorized" will be thrown.
+
+#### Before operations
+
+If a rule is only desired to run before operations, you will need to set `after` to an empty array:
+
+```gql
+type User
+  @authorization(
+    validate: [{ after: [], where: { node: { id: "$jwt.sub" } } }]
+  ) {
+  id: String!
+  name: String!
+}
+```
+
+Given the following GraphQL query:
+
+```gql
+{
+  users(where: { name: "Bob" }) {
+    id
+    name
+  }
+}
+```
+
+This will generate Cypher and parameters along the lines of the following:
+
+```cypher
+MATCH (this:User)
+WHERE this.name = $name
+AND apoc.util.validatePredicate(NOT (this.id = $jwt.sub), "Unauthorized", [])
+RETURN this { .id, .name }
+```
+
+```json
+{
+  "jwt": {
+    "sub": "123456"
+  },
+  "name": "Bob"
+}
+```
+
+If, following the user's filter application, the data to be returned does not satisfy the rule, the error "Unauthorized" will be thrown.
+
+#### After operations
+
+Like disabling `after` operations, you will need to do the same with `before` to run only after an operation:
+
+```gql
+type User
+  @authorization(
+    validate: [{ before: [], where: { node: { id: "$jwt.sub" } } }]
```

Comment:
I really don't like this approach of using empty `before` and `after` to define when an operation happens...
Not only seems like an ugly way of setting a flag, it is unintuitive as we are setting a before field to run the operations before.

Maybe I am missing something, but feels like this could use an enum or boolean instead :thinking:

Reply:
Yeah this is a sticking point, the issue is that the operations are different for `before` and `after`, so it's hard to unify them under a single boolean or something similar. I think this is an improvement to two completely different objects (`pre` and `post`), because I think most of the time people will want the same rule applied before and after operations.

Reply:
I think I don't follow on what the problem is, do you mind having a quick chat whenever you are free to check this?